### ISSUE TITLE
Accept multiple directories as extras

### ISSUE THREAD

**Issue by o01eg** [xi-editor/xi-editor#1212]
# Xi-Editor Issue Template
- [x] I have searched existing issues and could not find my issue.
- [x] I have studied the documentation.

## Expected vs Actual
Actually client can set only one extras dir.

In a discussion https://github.com/xi-frontend/xi-term/issues/110 it was mentioned to use XDG specification `XDG_DATA_DIRS` to place system-wide installed plugins, which is a collection of directories.

Can protocol be updated to allow client to set multiple directories for extras?

**Comment by Cogitri** [comment]
Hm, although I don't really oppose this wouldn't the current approach work for xi-term too? Setting config_dir (allowing users to put stuff in their `$HOME`) as it currently does and setting extra_dir to some other dir (e.g. `/usr/libexec/xi-term`, maybe via an env plugin) would do the trick too since `XDG_DATA_DIRS` most definitely isn't the right place for plugins. Another problem is predictability: How should `xi-core` choose what dir to use? The first one with a plugin in it? All of them? Xi also makes a theme cache, in which dir should that be in?

**Comment by o01eg** [comment]
> Hm, although I don't really oppose this wouldn't the current approach work for xi-term too? Setting config_dir (allowing users to put stuff in their $HOME) as it currently does and setting extra_dir to some other dir (e.g. /usr/libexec/xi-term, maybe via an env plugin) would do the trick too since XDG_DATA_DIRS most definitely isn't the right place for plugins.

Yes, but it's a more general issue to allow multiple extras directories.

> Another problem is predictability: How should xi-core choose what dir to use? The first one with a plugin in it? All of them?

I suppose xi-core could use all of them, `ConfigManager::get_plugin_paths` already returns multiple directories.

> Xi also makes a theme cache, in which dir should that be in?

Not sure about Windows, but on Linux it should use `XDG_CACHE_HOME`.

**Comment by Cogitri** [comment]
> Not sure about Windows, but on Linux it should use XDG_CACHE_HOME.

Which isn't available on all systems, let alone be handled by xi. I guess we could add a 3rd parameter to client_started

**Comment by cmyr** [comment]
My memory isn't totally clear but I believe xi _does_ allow the frontend to provide _two_ config directories, a 'system' one and a 'user' one. The idea being that the frontend can put bundled plugins and themes somewhere private, without the user stepping on them.

You specify this with the [`client_extras_dir` dir argument to `client_started`](https://github.com/xi-editor/xi-editor/blob/master/rust/core-lib/src/rpc.rs#L184). This may only be for plugins, I'm not totally sure, but it _is_ used in xi-mac so it should work. :)

**Comment by Cogitri** [comment]
> My memory isn't totally clear but I believe xi does allow the frontend to provide two config directories, a 'system' one and a 'user' one.

Yup, but the goal here was to use the `XDG` variables (basically standard dirs on linux for stuff like Downloads, Documents, Shared Files and so on), which includes multiple dirs (e.g. `$HOME/. local/share:/usr/local/share:/usr/share` so that the user can place the plugins in either of those dirs.

**Comment by cmyr** [comment]
I think in the particular case described above the current mechanism would work, though; the client just sends `$XDG_DATA_DIR/xi/plugins` as `client_extras_dir`. I don't have any ambition to support an arbitrary number of paths; I don't see a super compelling use case there.

**Comment by Cogitri** [comment]
> described above the current mechanism would work, though; the client just sends $XDG_DATA_DIR/xi/plugins as client_extras_dir

But XDG_DATA_DIR maps to multiple dirs, so that's not possible as of now. 

Anyway, I don't think it's worth it supporting multiple dirs either, it'd need some logic behind what path to prefer for conflicting files which I'd rather not have.
Would it be possible to just bake one dir into xi-term, @o01eg ? Tau does that to set a plugin dir.

**Comment by o01eg** [comment]
> Anyway, I don't think it's worth it supporting multiple dirs either, it'd need some logic behind what path to prefer for conflicting files which I'd rather not have.

XDG_DATA_DIRS is already preference-ordered as it is noted in specification.

> Would it be possible to just bake one dir into xi-term, @o01eg ? Tau does that to set a plugin dir.

It possible to take first directory from XDG_DATA_DIRS, I think.

**Comment by Cogitri** [comment]
> XDG_DATA_DIRS is already preference-ordered as it is noted in specification.

Yes, but frontends may specify dirs that aren't in the XDG standard too, so we'd need some logic behind this (e.g. with an enum which allows the front-end to select between different approaches for this). 

Anyway, in the short-term xi-term should just select one dir. Maybe you could draft a PR for xi-editor if you're interested in seeing this land?
